### PR TITLE
Automatically generate config documentation

### PR DESCRIPTION
**Pull request by petrutlucian94**
The bootstrap config documentation may get out of sync as a result of apiv1.BootstrapConfig changes.

To avoid this, we're going to automatically generate the documentation based on the apiv1.BootstrapConfig structure definition and docstrings. The same applies to the node join configuration.

This change adds the necessary tooling as part of the ``docgen`` package, which may be reused for other structures or projects. Note that nested structures are handled as well.

https://github.com/canonical/k8s-snap-api/pull/11 adds the necessary docstrings based on the current document.

### CODE REVIEW

**Comment by petrutlucian94** [comment]
The doc automation is ready, I need to add the missing docstrings to ``BootstrapConfig`` so that the tool can pick the description of each field.

**Comment by petrutlucian94** [comment]
``Documentation Spelling Check / spell-check (pull_request) `` is expected to fail until the other PR merges:

```
Detected changes that have not been committed to the repository
Please run 'cd src/k8s && make go.fmt go.doc'
```

We'll need to merge https://github.com/canonical/k8s-snap-api/pull/11, create a new release and update the ``k8s-snap`` requirements.

File: src/k8s/cmd/k8s/k8s_generate_docs.go [hunk 1]
```diff
@@ -15,11 +19,29 @@ func newGenerateDocsCmd(env cmdutil.ExecutionEnvironment) *cobra.Command {
 		Hidden: true,
 		Short:  "Generate markdown documentation",
 		Run: func(cmd *cobra.Command, args []string) {
-			if err := doc.GenMarkdownTree(cmd.Parent(), opts.outputDir); err != nil {
+			if err := doc.GenMarkdownTree(cmd.Parent(), opts.outputDir+"/commands"); err != nil {
```
Comment (bschimke95) [inline src/k8s/cmd/k8s/k8s_generate_docs.go:22]: nit: you may use `path.Join`

File: src/k8s/cmd/k8s/k8s_generate_docs.go [hunk 2]
```diff
@@ -15,11 +19,29 @@ func newGenerateDocsCmd(env cmdutil.ExecutionEnvironment) *cobra.Command {
...
+
+			bootstrap_doc, err := docgen.MarkdownFromJsonStruct(apiv1.BootstrapConfig{})
+			if err != nil {
+				cmd.PrintErrf("Error: Failed to generate markdown documentation for bootstrap configuration\n\n")
+				cmd.PrintErrf("Error: %v", err)
+				env.Exit(1)
+				return
+			}
+
+			bootstrap_doc_path := opts.outputDir + "/bootstrap_config.md"
```
Comment (bschimke95) [inline src/k8s/cmd/k8s/k8s_generate_docs.go:36]: same here

**Comment by petrutlucian94** [comment]
Screenshot: 
image

File: src/k8s/pkg/docgen/gomod.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+	return "", "", fmt.Errorf("could not find dependency %s in %s", packageName, goModPath)
+}
+
+// getProjectDir retrieves the full path of k8s-snap/src/k8s.
+// For simplicity, we assume that the executable is placed in
+// k8s-snap/src/k8s/bin/(static|dynamic).
+// This will mostly be used to generate the project documentation using
+// "make go.doc".
+func getProjectDir() (string, error) {
```
Comment (bschimke95) [inline src/k8s/pkg/docgen/gomod.go:74]: This introduces non-obvious dependencies on the project structure.
Does it make sense to pass the dir e.g. via flag or env?

**Review by bschimke95 — COMMENTED** [review]
Great work - just some minor questions/nits

File: src/k8s/pkg/docgen/gomod.go (hunk 3 above)
Comment (petrutlucian94) [reply src/k8s/pkg/docgen/gomod.go]: I went with a CLI parameter.

**Comment by bschimke95** [comment]
The Docs and Go unittests should now pass, right?

**Comment by petrutlucian94** [comment]
I'll address the ``go vet`` errors right away. However, the ``Documentation Spelling Check`` job reports a lot of unrelated errors:

```

Misspelled words:
<htmlcontent> ../_build/snap/index.html: html>body>div>aside>div>div>div>div>ul>li>ul>li.toctree-l2
--------------------------------------------------------------------------------
managment
--------------------------------------------------------------------------------

Misspelled words:
<htmlcontent> ../_build/snap/index.html: html>body>div>div>div>div>article>section>p
--------------------------------------------------------------------------------
performant
scalable
--------------------------------------------------------------------------------

Misspelled words:
<htmlcontent> ../_build/genindex/index.html: html>body>div>aside>div>div>div>div>ul>li>ul>li>ul>li.toctree-l3
--------------------------------------------------------------------------------
gapped
--------------------------------------------------------------------------------

Misspelled words:
<htmlcontent> ../_build/genindex/index.html: html>body>div>aside>div>div>div>div>p.caption
--------------------------------------------------------------------------------
WIP
--------------------------------------------------------------------------------

Misspelled words:
<htmlcontent> ../_build/genindex/index.html: html>body>div>aside>div>div>div>div>ul>li>ul>li.toctree-l2
--------------------------------------------------------------------------------
managment
--------------------------------------------------------------------------------

!!!Spelling check failed!!!
make: *** [Makefile:86: spelling] Error 1
Error: Process completed with exit code 2.
```

**Comment by bschimke95** [comment]
Could you create a follow-up PR that fixes the Spelling CI? Thanks!

**Comment by petrutlucian94** [comment]
I've fixed ``go vet``, currently looking at the spell check. We'll need to add some words to its ignore list and fix a few actual typos.

**Review by bschimke95 — APPROVED** [review]
LGTM

As discussed: Let's merge this PR and create a separate PR for the spelling fix.

**Pull request merged**